packages/utilities/src/FocusRectsProvider.tsx

```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { FocusRectsContext } from './useFocusRects';
+
+export type FocusRectsProviderParams = {
+  /**
+   * Ref to the root element that this is providing focus rects for.
+   */
+  providerRef: React.RefObject<HTMLElement>;
+
+  /**
+   * Indicates that this is the root of a layer, and should not inherit the providerRef from a parent context.
+   */
+  layerRoot?: boolean;
+};
+
+export const FocusRectsProvider: React.FC<FocusRectsProviderParams> = params => {
+  const { providerRef, layerRoot } = params;
+  const [registeredProviders] = React.useState<React.RefObject<HTMLElement>[]>([]);
+  const parentContext = React.useContext(FocusRectsContext);
+
+  // Inherit the parent context if it exists, unless this is a layer root.
+  // This allows the topmost provider element in the DOM tree to handle the focus events.
+  // Since layers are in a separate HTML tree from their parent, they shouldn't use the parent's providerRef.
+  const inheritParentContext = parentContext !== undefined && !layerRoot;
+
+  const context = React.useMemo(
+    () =>
+      inheritParentContext
+        ? undefined
+        : {
+            providerRef,
+            registeredProviders,
+            registerProvider: (ref: React.RefObject<HTMLElement>) => {
+              // Register this child provider with the current context, and any parent contexts
+              registeredProviders.push(ref);
+              parentContext?.registerProvider(ref);
+            },
+            unregisterProvider: (ref: React.RefObject<HTMLElement>) => {
```

Comment:
>unregisterProvider [](http://example.com/codeflow?start=12&length=18)

When a provider is unregistered it should clear any classname that it added to the classList? In the case where the handlers are hooked up to the body, you can get into a case where the handers get removed but the last value of the classname stays on the elment (e.g. ms-Fabric--isFocusVisible can hang around on the body even after the provider is cleaned up)

Reply:
Perhaps it should. However, this is working the same way that it did before my PR. I'd rather keep things working as similar as possible to how they were before this change, to avoid causing regressions in edge cases I'm not aware of.

In practice, the provider should be unmounted when the HTMLElement is unmounted, so stale class names shouldn't be hanging around.

This provider is not used to set the classname on the body--that's only done when there is no provider.

Reply:
Yep, I agree, I'll open a new bug for this

Reply:
I've opened https://github.com/microsoft/fluentui/issues/24419 to track that issue